File: app/src/main/java/io/mgba/Presenter/GamesPresenter.java
package io.mgba.Presenter;


import io.mgba.Presenter.Interfaces.IGamesPresenter;
import io.mgba.Data.Database.Game;
import io.mgba.Data.Platform;
import io.mgba.UI.Fragments.Interfaces.IGamesFragment;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class GamesPresenter implements IGamesPresenter {
    private final IGamesFragment<Game> view;
    private CompositeDisposable disposable = new CompositeDisposable();
    private Platform platform;

    public GamesPresenter(Platform platform, IGamesFragment<Game> view) {
        this.view = view;
        this.platform = platform;
    }

    @Override
    public void onDestroy() {
        disposable.dispose();
    }

    @Override
    public Platform getPlatform() {
        return platform;
    }

    @Override
    public void loadGames(io.mgba.UI.Activities.Interfaces.ILibrary databaseHelper) {
        disposable.add(databaseHelper.getLibraryService().prepareGames(platform)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(games -> {
                    view.swapContent(games);
                    view.showContent(games.size() > 0);
                }));
    }

    @Override
    public void onRefresh(io.mgba.UI.Activities.Interfaces.ILibrary iLibrary) {
        disposable.add(iLibrary.getLibraryService().reloadGames(platform)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(games -> {
                            view.swapContent(games);
                            view.stopRefreshing();
                        })
                );
    }

    @Override
    public Consumer<Game> getOnClick() {
        return view::handleItemClick;
    }
}

File: app/src/main/java/io/mgba/presenter/GamesPresenter.kt
package io.mgba.presenter

import io.mgba.presenter.interfaces.IGamesPresenter
import io.mgba.data.database.model.Game
import io.mgba.ui.fragments.interfaces.IGamesFragment
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class GamesPresenter(override val platform: Int, private val view: IGamesFragment<Game>) : IGamesPresenter {
    private val disposable = CompositeDisposable()

    override fun onClick(game: Game) {
        view.handleItemClick(game)
    }

    override fun onDestroy() {
        disposable.dispose()
    }

    override fun loadGames(databaseHelper: io.mgba.ui.activities.interfaces.ILibrary) {
        disposable.add(databaseHelper.libraryService.prepareGames(platform)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { games ->
                    view.swapContent(games)
                    view.showContent(games.size > 0)
                })
    }

    override fun onRefresh(iLibrary: io.mgba.ui.activities.interfaces.ILibrary) {
        disposable.add(iLibrary.libraryService.reloadGames(platform)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { games ->
                    view.swapContent(games)
                    view.stopRefreshing()
                }
        )
    }
}
